keps/sig-node/127-user-namespaces/README.md

```diff
@@ -0,0 +1,456 @@
+host files).
+
+#### Story 5
+
+As a cluster admin, I want to use different host UIDs/GIDs for pods running in
+the same node (whenever kernel/kube features allow it), so I can mitigate the
+impact a compromised pod can have on other pods and the node itself.
+
+### Notes/Constraints/Caveats
+
+#### UID space
+
+#### Volume support
+
+### Risks and Mitigations
+
+## Design Details
+
+Note: Names are preliminary yet, I'm using field names to simplify explanations.
+
+### Pod.spec changes
+
+The following changes will be done to the pod.spec:
+
+- `pod.spec.useHostUsers`: bool.
+If true or not present, uses the host user namespace (as today)
+If false, a new userns is created for the pod.
+This field will be used for phase 1, 2 and 3.
+
+- `pod.spec.securityContext.userns.pod2podIsolation`: Enum
+If enabled, we will make the userns mappings be non-overlapping as much as possible.
+This field will be used in phase 3.
+
+### Phases
+
+We propose to divide the work in 3 phases. Each phase makes this work with
+either more isolation or more workloads. When no support is yet added to handle
+some workload, a clear error will be shown.
+
+Please note the last sub-section here is a table with the summary of the changes
+proposed on each phase.
+
+#### Phase 1: pods "without" volumes
+
+This phase makes pods "without" volume work with user namespaces. This is
+activated via the bool `pod.spec.HostUsers` and works only when no volume or any
+combination of the following volume types are used by the pod:
+
+ - configmap
+ - secret
+ - downwardAPI
+ - emptyDir
+ - projected
+
+This list of volumes was chosen as they can't be used to share files with other
+pods.
+
+The mapping length will be 65535, mapping the range 0-65534 to the pod. This wide
+range makes sure most workloads will work fine.
+
+The mapping will be chosen by the kubelet, using a simple algorithm to give
+different pods in this category ("without" volumes) a non-overlapping mapping.
+Giving non-overlapping mappings generates the best isolation for pods.
+
+Furthermore, the node UID space of 2^32 can hold up to 2^16 pods each with a
+mapping length of 65535 (2^16-1) top. This imposes a limit of 65k pods per node,
+but that is not an issue we will hit in practice for a long time, if ever (today
+we run 110 pods per node by default).
+
+#### Phase 2: pods with volumes
+
+This phase makes user namespaces work in pods with volumes too. This is
+activated via the bool `pod.spec.HostUsers` too and pods fall into this mode if
+some other volume type than the ones listed for phase 1 is used. IOW, when phase
+2 is implemented, pods that use volume types not supported in phase 1, fall into
+the phase 2.
+
+All pods in this mode will use _the same_ mapping, chosen by the kubelet, with a
+length 65535, and mapping the range 0-65534 too. IOW, each pod will have its own user
+namespace, but they will map to _the same_ UIDs/GIDs in the host.
```

Comment:
Can we limit the pods sharing a volume to have the same mapping?

Reply:
Yes, this is what we propose in phase 3. Share the mappings per k8s namespace or service account. 

We propose it in phase 3 as it is not trivial. Besides the knobs like teaching [fsGroupChangePolicy](https://kubernetes.io/blog/2020/12/14/kubernetes-release-1.20-fsgroupchangepolicy-fsgrouppolicy/) about the effective UID used in the container (with userns RunAsUser is not the effective UID used) and similar things we need when we handle volumes (like in phase 2), the thing is that by sharing the mappings between nodes the mapping scope is now the cluster. Then if we have a mapping per k8s ns and each mapping is 64k, then we impose a maximum of 64k k8s namespaces to the cluster.

As Tim suggested that is completely possible today and he has seen things like that, we want to reduce the mapping length, so we don't cap so much the max namespaces a k8s cluster can have, when we explore that alternative. This is of course tricky, that is why we think is something to discuss in phase 3.

Other options that won't suffer from that max cap is using id mapped mounts, that we also suggest to explore in phase 3.

What do you think?